### PR TITLE
refactor for testing

### PR DESCRIPTION
**Pull request by justinnoah**
- Implement create_app to allow the test environment to modify config options (e.g. DABATASE_URI)
- Drop email_client.py as it was deprecated and thus unused
- Rename config DATABASE_URL to DATABASE_URI
- Clean up hard config imports
```python
# No more of this stuff
try:
   from config import CONFIG_OPTION
except:
    from configdist import CONFIG_OPTION
# Use app.config.get("CONFIG_OPTION")
# or current_app.config.get("CONFIG_OPTION") instead
```

### CODE REVIEW

File: _15thnight/api/account.py [hunk 1]
```diff
@@ -66,6 +61,7 @@ def forgot_password():
     """
     Send a password reset email.
     """
+    reset_token_life = current_app.config.get("RESET_TOKEN_LIFE", "")
```
Comment (nobane) [inline _15thnight/api/account.py]: Should these really need a default value? They should have a value set for this. If not maybe they should crash? Or maybe we should import configdist and pass that as the default value for this pattern.

File: _15thnight/__init__.py [hunk 2]
```diff
@@ -4,44 +4,63 @@
...
+
+    # Attempt to load config.py or fallback
+    try:
+        app.config.from_object('config')
+    except:
+        app.config.from_object('configdist')
+
+    # Additional configuration options (used by testing)
+    from _15thnight.database import init_db
+    ### OVERRIDE CONFIGURATION OPTIONS FOR TESTING HERE ###
```
Comment (nobane) [inline _15thnight/__init__.py]: This comment is kind of intense.

Comment (nobane) [reply _15thnight/__init__.py]: Also confusing, because I think you're just referring to line 35 here?

File: _15thnight/__init__.py [hunk 3]
```diff
@@ -4,44 +4,63 @@
...
 login_manager = LoginManager()
-login_manager.init_app(app)
-login_manager.login_view = 'login'
 
-csrf_protect.init_app(app)
 
-mailer.init_app(app)
+def create_app(cfg=dict()):
+    """
+    Setup an app with bluepirnts, extensions, and other necessary addtions.
```
Comment (nobane) [inline _15thnight/__init__.py]: Typo here `additions`

File: _15thnight/api/account.py (hunk 1 above)
Comment (justinnoah) [reply _15thnight/api/account.py]: For now, we can go with no default value as you suggest, allowing for a crash. Though in a future enhancement branch we probably should load configdist as a set of defaults. That's a really great idea.

File: _15thnight/__init__.py (hunk 2 above)
Comment (justinnoah) [reply _15thnight/__init__.py]: Intended to be noticed. Fixed up to be noticeable without such intensity for the next push.

File: _15thnight/__init__.py (hunk 3 above)
Comment (justinnoah) [reply _15thnight/__init__.py]: Also bluepirnts....

File: _15thnight/__init__.py [hunk 4]
```diff
@@ -4,44 +4,74 @@
...
+        # [0]: http://flask.pocoo.org/docs/0.11/api/#configuration
+        app.config.from_object('config')
+    except:
+        # Else keep the configdist defaults
+        pass
+
+    # Additional configuration options (used by testing)
+    from _15thnight.database import init_db
+
+    ### XXX: Override configuration options for testing here ###
```
Comment (nobane) [inline _15thnight/__init__.py]: Please remove the XXX from this commend, it is fine as-is without the XXX.

File: _15thnight/__init__.py [hunk 5]
```diff
@@ -4,44 +4,74 @@
...
+    app.config.update(cfg)
+
+    # XXX: must use app.db_session or current_app.db_session because of this
+    app.db_session = init_db(app.config.get("DATABASE_URI"))
+    # app.secret_key is used by Flask-WTF
+    app.secret_key = app.config['SECRET_KEY']
+    # TODO: Comment by @spectralsun needed
+    # XXX: why does this need to be tied to the app?
```
Comment (nobane) [inline _15thnight/__init__.py]: This hooks into the jsonify that flask uses to turn objects into json.

File: _15thnight/__init__.py [hunk 6]
```diff
@@ -4,44 +4,74 @@
...
+        # Else keep the configdist defaults
+        pass
+
+    # Additional configuration options (used by testing)
+    from _15thnight.database import init_db
+
+    ### XXX: Override configuration options for testing here ###
+    app.config.update(cfg)
+
+    # XXX: must use app.db_session or current_app.db_session because of this
```
Comment (nobane) [inline _15thnight/__init__.py]: No need for a comment here.

File: _15thnight/__init__.py [hunk 7]
```diff
@@ -4,44 +4,74 @@
...
+    app.db_session = init_db(app.config.get("DATABASE_URI"))
+    # app.secret_key is used by Flask-WTF
```
Comment (nobane) [inline _15thnight/__init__.py:45]: Probably don't need to comment this.

File: _15thnight/__init__.py [hunk 8]
```diff
@@ -91,4 +121,4 @@ def logout():
 @app.route('/health')
```
Comment (nobane) [inline _15thnight/__init__.py]: We can delete this route.

File: _15thnight/__init__.py [hunk 9]
```diff
@@ -4,44 +4,74 @@
...
 from _15thnight.util import ExtensibleJSONEncoder
 
 
-app = Flask(__name__)
-
-try:
-    app.config.from_object('config')
-except:
-    app.config.from_object('configdist')
+login_manager = LoginManager()
```
Comment (nobane) [inline _15thnight/__init__.py:13]: Move this line and all the lines below create_app that are using the `@app` annotation into the `create_app` function please. We will want to test that stuff eventually, and for cleanliness, it should all be in `create_app`.

File: _15thnight/database.py [hunk 10]
```diff
@@ -1,20 +1,11 @@
 import os
 
+from flask import g
```
Comment (nobane) [inline _15thnight/database.py]: NOOOOOOOOOOOOOO `g`

File: _15thnight/database.py [hunk 11]
```diff
@@ -35,29 +26,32 @@ def get_by_ids(cls, id_list):
...
+def init_db(db_uri=""):
+    # Create database engine
+    engine = create_engine(db_uri, convert_unicode=True)
 
-# Create our scoped database session and bind to the engine
-db_session = scoped_session(sessionmaker(autocommit=False,
+    # Create our scoped database session and bind to the engine
+    db_session = scoped_session(sessionmaker(autocommit=False,
                                          autoflush=False,
                                          bind=engine))
```
Comment (nobane) [inline _15thnight/database.py]: Fix up the indenting here please

File: _15thnight/__init__.py (hunk 9 above)
Comment (justinnoah) [reply _15thnight/__init__.py:13]: It cannot be. Order of operations requires this with the current setup.

File: _15thnight/database.py (hunk 10 above)
Comment (justinnoah) [reply _15thnight/database.py]: Extraneous import, it's not being used. I'll remove it.

File: _15thnight/__init__.py (hunk 6 above)
Comment (justinnoah) [reply _15thnight/__init__.py]: Removed the XXX, though code documentation is important. Having too many comments is better than none. IMHO

**Comment by justinnoah** [comment]
@spectralsun conflicts resolved

**Pull request closed without merging**